File: c/letCombPh.c
/* Given a string containing digits from 2-9 inclusive, return all possible
 * letter combinations that the number could represent.
 * Return the answer in any order.
 * */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define PRINT(a,b)                 do {                            \
    if(a==NULL) printf("[]\n-------------------\n");              \
    else{                                                         \
    int i;                                                         \
    printf("[");                                                   \
    for(i=0;i<b-1;i++)                                          \
      printf("%s,",a[i]);                                            \
    printf("%s]\n-----------------\n",a[i]); }}                      \
  while(0)

static char *NUMMAP[] = {
  "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"
};

void innerComb(char **, int, int *, int l, int *, int *,char *[]);

char ** letterCombinations(char * digits, int *returnSize)
{
  char *p=digits, *q, **res;
  int prd = 1,n=0,i,points=0;

  int l = strlen(digits), *lens = (int *)malloc(l*sizeof(int)) , *pos = (int *)malloc(l*sizeof(int));
  if(l ==0) {
    *returnSize = 0;
    return NULL;
}
  char **locs = (char **)malloc(l*sizeof(char*));

  for(q = p, i = 0; *q;q++,i++ ) {
    lens[i] = strlen(NUMMAP[*q -'0']);
    prd *= lens[i];
    pos[i] = 0;
    locs[i] = strdup(NUMMAP[*q - '0']);
    /* locs[i] = NUMMAP[*q - '0']; */
  }

  res = (char **)malloc(prd*sizeof(char*));
  *returnSize = prd;

  for(i=0;i<prd;i++)
    res[i] = (char *)malloc((l+1)*sizeof(char));

  innerComb(res,n, lens, l, pos, &points, locs);
  return res;

}

void innerComb(char ** res, int n, int *lens, int l, int *pos, int *points, char *LOC[])
{
  int i,ii, l1, l2;
  char *q;
  n++;
  for(i=0;i<lens[ n - 1];i++) {
    if (n % l == 0){
      pos[n - 1] = i;
      l1 = *points / 3;
      l2 = *points % 3;
      q = res[3*l1+l2];
      for (ii=0;ii<l;*q++ = LOC[ii][pos[ii]],ii++);
      *q = '\0';
      *points += 1;

    }
    else {
      pos[n -1] = i;
      innerComb(res, n, lens, l, pos, points,LOC);
    }

  }

}

#if 0
char map[10][5] = {
					{},
					{},
					{'a','b','c'},
					{'d','e','f'},
					{'g','h','i'},
					{'j','k','l'},
					{'m','n','o'},
					{'p','q','r','s'},
					{'t','u','v'},
					{'w','x','y','z'}
				  };

int result_index;
char **result;

void helper(int index, char *tempstr, char *digits) {
    if(digits[index] == '\0') {
        tempstr[index] = '\0';
        result[result_index] = (char*)malloc(sizeof(char) * (strlen(digits) + 1));
        strcpy(result[result_index], tempstr);
        result_index++;
    }else {
        for(int i = 0 ; i < strlen(map[(digits[index] - '0')]) ; i++) {
            tempstr[index] = map[(digits[index] - '0')][i];
            helper(index + 1, tempstr, digits);
        }
    }
}
char **letterCombinations(char *digits, int *returnSize) {
    result_index = 0;

    if(strlen(digits) != 0) {
        (*returnSize) = 1;
        for(int i = 0 ; i < strlen(digits) ; i++ )
            (*returnSize) = (*returnSize) * strlen(map[(digits[i] - '0')]);
    }
    else
		(*returnSize) = 0;

    result = (char**)malloc(sizeof(char*) * (*returnSize));

    char tempstr[5] = {0};
    if((*returnSize) > 0)
        helper(0, tempstr, digits);

    return result;
}
#endif

int main()
{
  char **ret;
  int retSize;

  char *digits = "23";
  ret = letterCombinations(digits,&retSize);
  PRINT(ret,retSize);

  char *digits1 = "27";
  ret = letterCombinations(digits1,&retSize);
  PRINT(ret,retSize);

  char *digits2 = "2";
  ret = letterCombinations(digits2,&retSize);
  PRINT(ret,retSize);

  char *digits4 = "";
  ret = letterCombinations(digits4,&retSize);
  PRINT(ret,retSize);

  char *digits3 = "2476";
  ret = letterCombinations(digits3,&retSize);
  PRINT(ret,retSize);

  return 0;
}

File: python/letCombPh.py
"""
/* Given a string containing digits from 2-9 inclusive, return all possible
 * letter combinations that the number could represent.
 * Return the answer in any order.
 * */
"""

NUM_MAP = {
    2:"abc",
    3:"def",
    4:"ghi",
    5:"jkl",
    6:"mno",
    7:"pqrs",
    8:"tuv",
    9:"wxyz"
}

def lettComb(string):
    nos = []
    res = []
    pos = []
    lens = []

    points = 0
    n = 0
    prd = 1
    for ii,s in enumerate(string):
        i = int(s)
        nos.append(NUM_MAP[i])
        n += 1
        pos.append(0)
        lens.append(len(NUM_MAP[i]))
        prd *= lens[ii]

    print(f"Nos are --{nos}")
    for _ in range(prd):
        res.append([-1]*n)

    def innerComb(nn,ind=0):
        """
        @param: nn is used to iterate over the numbers
        @param: ind is used to track which of the inner array indexing into
        @param: pos is used to track position from which to iterate
        """
        nonlocal points
        # nonlocal pos

        # if points == 3**n:
            # return              # we're done
        ind += 1
        # print(f"Points={points}--ind={ind}--pos={pos}")
        # print("-"*30)

        for i in range(lens[ind - 1]):
            if (ind) % n ==0: # Time to add
                pos[ind - 1] = i
                loc1 = points // 3
                loc2 = points % 3
                for ii in range(n):
                    # print(f"\t\tloc1={loc1}--loc2={loc2}--ii={ii}")
                    res[3*loc1+loc2 ][ii] = nos[ii][pos[ii]]
                    # print(f"\t\tResult is {res}")
                points += 1
            else:
                pos[ind - 1] = i
                innerComb(nn,ind)




    innerComb(n)
    print("----------------xxxxxxxxxx--------------------------")
    return ["".join(aa) for aa in res]

def iterletComb(n=3):
    nos = []
    res = []
    points = []

    for i in range(n):
        nos.append(list(range(3*i,3*i+3)))
        points.append(0)

    for i in range(3**n):
        res.append([-1,-1,-1])

    points = 0
    for ii in range(3):
        for jj in range(3):
            for kk in range(3):
                loc1 =  points // 3
                loc2 = points % 3
                res[3*loc1+loc2 ][0] = nos[0][ii]
                res[3*loc1+loc2][1] = nos[1][jj]
                res[3*loc1+loc2][2] = nos[2][kk]
                points += 1


    return res



if __name__ == '__main__':

    res = iterletComb()
    res = lettComb("23")
    print(res,len(res),sep='\n')
    print("*"*50)

    res = lettComb("27")
    print(res,len(res),sep='\n')
    print("*"*50)

    res = lettComb("2")
    print(res,len(res),sep='\n')
    print("*"*50)

    res = lettComb("2476")
    print(res,len(res),sep='\n')
    print("*"*50)
